tests/pipeline_operations_test.py

```diff
@@ -385,6 +386,254 @@ def test_local_flat_map(self):
             [("a", 1), ("a", 2), ("a", 3), ("a", 4), ("b", 5), ("b", 6),
              ("b", 7), ("b", 8)])
 
+    def test_local_group_by_key(self):
+        some_dict = [("cheese", "brie"), ("bread", "sourdough"),
+                     ("cheese", "swiss")]
+
+        self.assertEqual(list(self.ops.group_by_key(some_dict)),
+                         [("cheese", ["brie", "swiss"]),
+                          ("bread", ["sourdough"])])
+
+class MultiProcLocalPipelineOperationsTest(unittest.TestCase):
+    @staticmethod
+    def partition_extract(x): return x[1]
+
+    @staticmethod
+    def privacy_id_extract(x): return x[0]
+
+    @staticmethod
+    def value_extract(x): return x[2]
+
+    @classmethod
+    def setUpClass(cls):
+        cls.ops = MultiProcLocalPipelineOperations(n_jobs=1)
+        cls.data_extractors = DataExtractors(
+            partition_extractor=cls.partition_extract,
+            privacy_id_extractor=cls.privacy_id_extract,
+            value_extractor=cls.value_extract)
+
+    def sortDataset(self, data):
+        if isinstance(data, (str, int, float)):
+            return data
+        if isinstance(data, tuple):
+            return tuple(self.sortDataset(x) for x in data)
+        data = list(data)
+        def sortfn(x):
+            # make everything a sortable thing!
+            if isinstance(x, list):
+                return self.sortDataset(tuple(x))
+            return x
+        return sorted(data, key=sortfn)
+
+    def assertDatasetsEqual(self, first, second, toplevel=True):
```

Comment:
it looks like we need to compare first and second as multisets. It looks like in Python [Counter](https://docs.python.org/3/library/collections.html#collections.Counter) is best approach for multisets. Wouldn't it work?

Reply:
This is a somewhat generic implementation of comparing two datasets that can hold arbitrary objects. Value types like `int, str, float` are compared as is, tuples are compared by value as well (unless they hold only 2 elements in which case we assume it might be a key-value pair where the value is another dataset), and lists are treated as multisets.  
[`collections.Counter`](https://docs.python.org/3/library/collections.html#collections.Counter) won't work in this case because some of the types I'm handling are **mutable** (lists for example) and hence cannot serve as dictionary keys.

Reply:
In essence, I implicitly define `Dataset`:
```python
ValueType = int | str | float
DatasetType = ValueType | Tuple[ValueType, 'DatasetType'] | Tuple['DatasetType'] | List['DatasetType']
```
Where:  
1. `ValueType` is compared by value directly.
2. `Tuple[ValueType, 'DatasetType']` compares the first element by value, and the second element using `DatasetType` rules.
3. `Tuple['DatasetType']` compares **ordered elementwise** elements, each element is assumed `DatasetType`.
4. `List['DatasetType']` compares the elements in a "multiset" fashion, where each element is assumed `DatasetType`.

Reply:
Thanks it makes sense